File: Bypyhton/有效括号.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2 == 1:
            return False
        pairs = {')': '(', ']': '[', '}': '{'}
        stack = []
        for ch in s:
            if ch in '([{':
                stack.append(ch)
            else:
                if not stack or stack.pop() != pairs[ch]:
                    return False
        return not stack

File: ByCpp/有效括号.cpp
#include <string>
#include <stack>

using namespace std;

bool isValid(string s) {
    if (s.size() % 2 == 1) return false;
    stack<char> st;
    for (char c : s) {
        if (c == '(' || c == '[' || c == '{') {
            st.push(c);
        } else {
            if (st.empty()) return false;
            char left = st.top(); st.pop();
            if ((c == ')' && left != '(') ||
                (c == ']' && left != '[') ||
                (c == '}' && left != '{')) {
                return false;
            }
        }
    }
    return st.empty();
}
